### PR TITLE
i#1621 AArch64 cc opt: Use analyze_callee_regs_usage for GPRs.

### PR DESCRIPTION
**Pull request by fhahn**
This patch implements analyze_callee_regs_usage to analyze the usage of
GPRs in the callee. insert_push_all_registers and
insert_pop_all_registers do not save and restore unused registers.

cleanall_opt=1 has limited support now and is the default level for
AArch64.

Based on work by Kevin Zhou and Edmund Grimley-Evans.

### CODE REVIEW

**Comment by fhahn** [comment]
My plan is to add clean call optimizations step by step in separate pull requests, starting with analyzing GPRs. I'll create additional pull requests adding more analysis and optimizations (SIMD register analysis, callee saved reg analysis, cc inlining). Please let me know in case you prefer a different sizing of the patches.

**Comment by zhaoqin** [comment]
The code looks overlaps with the one with helper function, should I review it after you update this?

**Comment by fhahn** [comment]
yes, I'll update this PR once the helper functions is merged and ping you :)

**Comment by fhahn** [comment]
I've updated the patch now. I also realized that a type slipped in with the latest commit I merged, which is also fixed here: https://github.com/DynamoRIO/dynamorio/pull/2342/commits/c6cd25b4e2e807cef7a131589c044dd3d082436a

File: core/arch/aarch64/clean_call_opt.c [hunk 1]
```diff
@@ -34,25 +34,74 @@
...
+        reg_id_t reg = regparms[i];
+        if (!ci->reg_used[reg - DR_REG_START_GPR]) {
+            LOG(THREAD, LOG_CLEANCALL, 2,
+                "CLEANCALL: callee "PFX" uses REG %s for arg passing\n",
+                ci->start, reg_names[reg]);
+            ci->reg_used[reg - DR_REG_START_GPR] = true;
+            callee_info_reserve_slot(ci, SLOT_REG, reg);
+        }
+    }
+    /* FIXME i#1621: the following checks are still missing:
```
Comment (zhaoqin) [inline core/arch/aarch64/clean_call_opt.c]: ok, the code is very similar to (not exactly the same) a subset of analyze_callee_regs_usage in arch/x86/clean_call_opt.c.
If there are too many duplicated code and easy to extract in the future, maybe we want to have separate helper function to share the code.

Comment (fhahn) [inline core/arch/aarch64/clean_call_opt.c]: I agree that's something we should look into in the future

**Pull request merged**